common/src/main/java/io/crate/types/TypeSignature.java

```diff
@@ -0,0 +1,175 @@
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.types;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+import static java.lang.String.format;
+
+public class TypeSignature {
+
+    /**
+     * Creates a type signature out of the given signature string.
+     * A signature type string may contain parameters inside parenthesis:
+     * <p>
+     *   base_type_name(parameter [, parameter])
+     * </p>
+     *
+     * Custom parameterized type handling must also be supported by {@link #createType()}.
+     *
+     * Some examples:
+     * <p>
+     *      integer
+     *      array(integer)
+     *      array(E)
+     *      object(text, integer)
+     *      object(text, V)
+     * <p>
+     */
+    public static TypeSignature parseTypeSignature(String signature) {
+        if (!signature.contains("(")) {
+            return new TypeSignature(signature);
+        }
+
+        String baseName = null;
+        List<TypeSignature> parameters = new ArrayList<>();
+        int parameterStart = -1;
+        int bracketCount = 0;
+
+        for (int i = 0; i < signature.length(); i++) {
+            char c = signature.charAt(i);
+            if (c == '(') {
+                if (bracketCount == 0) {
+                    assert baseName == null : "Expected baseName to be null";
+                    baseName = signature.substring(0, i);
+                    parameterStart = i + 1;
+                }
+                bracketCount++;
+            } else if (c == ')') {
+                bracketCount--;
+                if (bracketCount == 0) {
+                    assert parameterStart >= 0 : "Expected parameter start to be >= 0";
+                    parameters.add(parseTypeSignatureParameter(signature, parameterStart, i));
+                    parameterStart = i + 1;
+                    if (i == signature.length() - 1) {
+                        return new TypeSignature(baseName, parameters);
+                    }
+                }
+            } else if (c == ',') {
+                if (bracketCount == 1) {
```

Comment:
Does/ should this work for nested cases?

Reply:
Yes it does, I've added a test case.